community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexUpdatesValidator.java

```diff
@@ -0,0 +1,164 @@
+import org.neo4j.helpers.collection.Visitor;
+import org.neo4j.kernel.api.index.NodePropertyUpdate;
+import org.neo4j.kernel.impl.api.TransactionApplicationMode;
+import org.neo4j.kernel.impl.store.NeoStore;
+import org.neo4j.kernel.impl.store.NodeStore;
+import org.neo4j.kernel.impl.store.PropertyStore;
+import org.neo4j.kernel.impl.store.record.PropertyRecord;
+import org.neo4j.kernel.impl.transaction.TransactionRepresentation;
+import org.neo4j.kernel.impl.transaction.command.Command;
+import org.neo4j.kernel.impl.transaction.command.NeoCommandHandler;
+import org.neo4j.kernel.impl.transaction.state.LazyIndexUpdates;
+import org.neo4j.kernel.impl.transaction.state.PropertyLoader;
+
+import static org.neo4j.kernel.impl.transaction.command.Command.NodeCommand;
+import static org.neo4j.kernel.impl.transaction.command.Command.PropertyCommand;
+
+public class IndexUpdatesValidator
+{
+    private final NodeStore nodeStore;
+    private final PropertyStore propertyStore;
+    private final PropertyLoader propertyLoader;
+    private final IndexingService indexing;
+
+    public IndexUpdatesValidator( NeoStore neoStore, PropertyLoader propertyLoader, IndexingService indexing )
+    {
+        this.nodeStore = neoStore.getNodeStore();
+        this.propertyStore = neoStore.getPropertyStore();
+        this.propertyLoader = propertyLoader;
+        this.indexing = indexing;
+    }
+
+    public ValidatedIndexUpdates validate( TransactionRepresentation transaction, TransactionApplicationMode mode )
+            throws IOException
+    {
+        NodePropertyCommandsExtractor extractor = new NodePropertyCommandsExtractor();
+        transaction.accept( extractor );
+
+        if ( extractor.noCommandsExtracted() )
+        {
+            return ValidatedIndexUpdates.NONE;
+        }
+
+        if ( mode == TransactionApplicationMode.RECOVERY )
+        {
+            return newValidatedRecoveredUpdates( extractor.changedNodeIds(), indexing );
+        }
+
+        Iterable<NodePropertyUpdate> updates = new LazyIndexUpdates( nodeStore, propertyStore, propertyLoader,
+                extractor.propertyCommandsByNodeIds, extractor.nodeCommandsById );
+
+        return indexing.validate( updates );
+    }
+
+    private static ValidatedIndexUpdates newValidatedRecoveredUpdates( final PrimitiveLongSet nodeIds,
+            final IndexingService indexing )
+    {
+        return new ValidatedIndexUpdates()
+        {
+            @Override
+            public void flush()
+            {
+                indexing.addRecoveredNodeIds( nodeIds );
+            }
+
+            @Override
+            public void close()
+            {
+            }
+        };
+    }
+
+    private static class NodePropertyCommandsExtractor
+            extends NeoCommandHandler.Adapter implements Visitor<Command,IOException>
+    {
+        final PrimitiveLongObjectMap<NodeCommand> nodeCommandsById = Primitive.longObjectMap();
```

Comment:
@tinwelint, I'm not completely sure that it is possible to specify correct initial sizes here as we do not even know how many commands are there in transaction state. We can either expose additional method like `TransactionRepresentation#size` that will return number of commands and estimate size based on it (say TR.size / 3) or we can simply default to 16 just to make initial capacity smaller. What do you think?